File: pset6/sentimental-credit/credit.py
# TODO
import sys


def main():
    number = input("Number: ")
    length = len(str(number))

    if length != 13 and length != 15 and length != 16:
        print("INVALID")
        sys.exit()

    i = length - 2
    sum = 0
    while (i >= 0):
        if int(number[i]) < 5:
            sum += int(number[i]) * 2
        else:
            sum += 1 + (int(number[i]) * 2) % 10
        i -= 2

    i = length - 1
    while (i >= 0):
        sum += int(number[i])
        i -= 2

    check = int(number[0]) * 10 + int(number[1])
    if sum % 10 != 0:
        print("INVALID")
    else:
        if int(number[0]) == 4:
            print("VISA")
        elif check in [34, 37]:
            print("AMEX")
        elif 51 <= check <= 55:
            print("MASTERCARD")
        else:
            print("INVALID")


if __name__ == "__main__":
    main()

File: workspace/pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#define LENGTH 20

int main(void)
{
    long num, tmp_num;
    int digits = 1;
    int a[LENGTH] = {0};
    int sum = 0;

    do
    {
        num = get_long("Number: ");
    }
    while (num <= 0 || num >= 1e17);

    // 将每一位数储存在数组中
    for (digits = 0, tmp_num = num; tmp_num > 1; digits++)
    {
        a[digits] = tmp_num % 10;
        tmp_num /= 10;
    }

    // 隔一位乘2并求和
    for (int i = 1; i < digits; i += 2)
    {
        if (a[i] < 5)
        {
            sum += a[i] * 2;
        }
        else
        {
            sum += a[i] / 5 + (a[i] * 2) % 10;
        }
    }

    // 与未乘的数加起来
    for (int i = 0; i < digits; i += 2)
    {
        sum += a[i];
    }
    
    //有效性判断
    if (sum % 10 != 0)
    {
        printf("INVALID\n");
    }
    else
    {
        num /= 1e12;
        if (num >= 4 && num < 5)
        {
            printf("VISA\n");
        }
        else if (num >= 340 && num < 350)
        {
            printf("AMEX\n");
        }
        else if (num >= 370 && num < 380)
        {
            printf("AMEX\n");
        }
        else if (num >= 4000 && num < 5000)
        {
            printf("VISA\n");
        }
        else if (num >= 5100 && num < 5600)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
}
